apps/els_lsp/src/els_code_action_provider.erl

```diff
@@ -29,79 +29,133 @@ handle_request({document_codeaction, Params}, State) ->
 %% Internal Functions
 %%==============================================================================
 
-
 %% @doc Result: `(Command | CodeAction)[] | null'
 -spec code_actions(uri(), range(), code_action_context()) -> [map()].
-code_actions(Uri, _Range, Context) ->
-  #{ <<"diagnostics">> := Diagnostics } = Context,
-  Actions0 = [ make_code_action(Uri, D) || D <- Diagnostics],
-  Actions = lists:flatten(Actions0),
-  Actions.
-
-%% @doc Note: if the start and end line of the range are the same, the line
-%% is simply added.
--spec replace_lines_action(uri(), binary(), binary(), binary(), range())
-                          -> map().
-replace_lines_action(Uri, Title, Kind, Lines, Range) ->
-  #{ <<"start">> := #{ <<"character">> := _StartCol
-                     , <<"line">>      := StartLine }
-   , <<"end">>   := #{ <<"character">> := _EndCol
-                     , <<"line">>      := EndLine }
-   } = Range,
+code_actions(Uri, _Range, #{<<"diagnostics">> := Diagnostics}) ->
+  lists:flatten([make_code_action(Uri, D) || D <- Diagnostics]).
+
+-spec make_code_action(uri(), map()) -> [map()].
+make_code_action(Uri, #{<<"message">> := Message, <<"range">> := Range}) ->
+  make_code_action(
+    [ {"function (.*) is unused", fun action_export_function/3}
+    , {"variable '(.*)' is unused", fun action_ignore_variable/3}
+    , {"variable '(.*)' is unbound", fun action_suggest_variable/3}
+    , {"Module name '(.*)' does not match file name '(.*)'",
+       fun action_fix_module_name/3}
+    ], Uri, Range, Message).
+
+-spec make_code_action([{string(), Fun}], uri(), range(), binary()) -> [map()]
+          when Fun :: fun((uri(), range(), [binary()]) -> [map()]).
+make_code_action([], _Uri, _Range, _Message) ->
+  [];
+make_code_action([{RE, Fun}|Rest], Uri, Range, Message) ->
+  Actions = case re:run(Message, RE, [{capture, all_but_first, binary}]) of
+              {match, Matches} ->
+                Fun(Uri, Range, Matches);
+              nomatch ->
+                []
+            end,
+  Actions ++ make_code_action(Rest, Uri, Range, Message).
+
+-spec action_export_function(uri(), range(), [binary()]) -> [map()].
+action_export_function(Uri, _Range, [UnusedFun]) ->
```

Comment:
Matter of taste, but we could move the actual business logic to a separate module (say `els_code_actions`).